src/pudl/analysis/record_linkage/link_cross_year.py

```diff
@@ -0,0 +1,273 @@
+    min_samples: int = 2
+
+
+@op
+def cluster_records_dbscan(
+    config: DBSCANConfig, distance_matrix: DistanceMatrix, original_df: pd.DataFrame
+) -> pd.DataFrame:
+    """Generate initial IDs using DBSCAN algorithm."""
+    # DBSCAN is very efficient when passed a sparse radius neighbor graph
+    neighbor_computer = NearestNeighbors(radius=config.eps, metric="precomputed")
+    neighbor_computer.fit(distance_matrix.distance_matrix)
+    neighbor_graph = neighbor_computer.radius_neighbors_graph(mode="distance")
+
+    # Classify records
+    classifier = DBSCAN(metric="precomputed", eps=config.eps, min_samples=2)
+
+    # Create dataframe containing only report year and label columns
+    return pd.DataFrame(
+        {
+            "report_year": original_df.loc[:, "report_year"],
+            "record_label": classifier.fit_predict(neighbor_graph),
+        }
+    )
+
+
+class SplitClustersConfig(Config):
+    """Configuration for AgglomerativeClustering used to split overmerged clusters."""
+
+    #: See :class:`sklearn.cluster.AgglomerativeClustering` for details.
+    distance_threshold: float = 0.3
+
+
+@op
+def split_clusters(
+    config: SplitClustersConfig,
+    distance_matrix: DistanceMatrix,
+    id_year_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """Split clusters with multiple records from same report_year.
+
+    DBSCAN will sometimes match records from the same report year, which breaks the
+    assumption that there should only be one record for each entity from a single
+    report year. To fix this, agglomerative clustering will be applied to each
+    such cluster. Agglomerative clustering could replace DBSCAN in the initial linkage
+    step to avoid these matches in the first place, however, it is very inneficient on
+    a large number of records, so applying to smaller sets of overmerged records is
+    much faster and uses much less memory.
+    """
+
+    def _generate_cluster_ids(max_cluster_id: int) -> int:
+        """Get new unique cluster id."""
+        while True:
+            max_cluster_id += 1
+            yield max_cluster_id
+
+    cluster_id_generator = _generate_cluster_ids(id_year_df.record_label.max())
+    classifier = AgglomerativeClustering(
+        metric="precomputed",
+        linkage="average",
+        distance_threshold=config.distance_threshold,
+        n_clusters=None,
+    )
+    duplicated_ids = id_year_df.loc[
+        id_year_df.duplicated(subset=["report_year", "record_label"]),
+        "record_label",
+    ]
+
+    for duplicated_id in duplicated_ids.unique():
+        # IDs of -1 will be handled seperately
+        if duplicated_id == -1:
+            continue
+
+        cluster_inds = id_year_df[
+            id_year_df.record_label == duplicated_id
+        ].index.to_numpy()
+        cluster_distances = get_cluster_distance_matrix(
+            distance_matrix.distance_matrix, cluster_inds
+        )
+
+        new_labels = classifier.fit_predict(cluster_distances)
```

Comment:
Yeah this will split clusters with a duplicate year record into two or more new clusters. The benefit of doing it like this is that after the `dbscan` step, there are many clusters that seem to contain two distinct groupings within a single cluster, and this efficiently separates them.